compiler/src/lnntoamm/Event.ts

```diff
@@ -0,0 +1,99 @@
+import { LPNode } from "../lp";
+import Output from "./Amm";
+import Fn from './Fn';
+import Scope from "./Scope";
+import Type, { Builtin } from "./Types";
+import { genName, TODO } from "./util";
+
+var allEvents: Event[] = [];
+
+export default class Event {
+  ammName: string
+  name: string
+  eventTy: Type
+  // if it's a single fn
+  handlers: Array<Fn | Fn[]>
+  runtimeDefined: boolean
+
+  static get allEvents(): Event[] {
+    return allEvents;
+  }
+
+  constructor(
+    name: string,
+    eventTy: Type,
+    handlers: Array<Fn | Fn[]> = [],
+    runtimeDefined: boolean = false,
+  ) {
+    this.name = name;
+    this.eventTy = eventTy;
+    this.handlers = handlers;
+    if (allEvents.some(event => event.name === this.name)) {
+      this.ammName = genName();
+    } else {
+      this.ammName = this.name;
+    }
+    this.runtimeDefined = runtimeDefined;
+    allEvents.push(this);
+  }
+
+  static fromAst(ast: LPNode, scope: Scope): Event {
+    const name = ast.get('variable').t.trim();
+    const tyName = ast.get('fulltypename').t.trim();
+    const ty = scope.get(tyName);
+    if (ty === null) {
+      throw new Error(`Could not find specified type: ${tyName}`);
+    } else if (!(ty instanceof Type)) {
+      throw new Error(`${tyName} is not a type`);
+    }
+    return new Event(name, ty);
+  }
+
+  compile(amm: Output) {
+    if (this.runtimeDefined === false) {
+      amm.global('event', this.eventTy as Builtin, this.ammName);
+    }
+    for (let handler of this.handlers) {
+      if (handler instanceof Array) {
+        return TODO('Event handler selection');
+      }
+      if (!(handler instanceof Fn)) {
+        throw new Error('Too many possible event handlers');
+      }
+      // let [varConstraints, retConstraints] = handler.constraints([this.eventTy]);
+      // for (let {dec, constraints} of varConstraints) {
+      //   if (!constraints.every(con => dec.ty.compatibleWithConstraint(con))) {
+      //     throw new Error(`failed to type-check: declaration ${dec}`);
+      //   }
+      // }
+      // retConstraints.push(opcodes().get('void') as Type);
+      // really tsc????? you're gonna complain that handler could be an array
+      // here but not above? wat?
+      // if (!retConstraints.every(con => (handler as Fn).retTy.compatibleWithConstraint(con))) {
+      //   throw new Error(`failed to type-check: event handler ${handler}`);
+      // }
+      handler.asHandler(amm, this.ammName);
+    }
+  }
```

Comment:
Two general comments here:

1. You're allowed to have mutliple handlers registered to a single event. the runtimes will copy the payload to all of them an let them run, so it's not as much of a selection as an actual array of handlers to emit.
2. Is this commented code intended to be deleted before this is landed or no? If not, can a small `TODO` comment on top explaining what it's for be added?

Reply:
the comments are deleted code, I'll go ahead and clean up.

RE: multiple handlers, good to know! I'll fix that

Reply:
added a comment to clarify that the handlers have to be filtered to ensure that they accept the event type and return `void`